crypto/cmp/cmp_util.c

```diff
@@ -0,0 +1,422 @@
+ * when MSG_SIG_ALG is utilized.
+ *
+ * NOTE: This allocates a stack and increments the reference count of each cert,
+ * so when not needed any more the stack and all its elements should be freed.
+ * NOTE: in case there is more than one possibility for the chain,
+ * OpenSSL seems to take the first one, check X509_verify_cert() for details.
+ *
+ * returns a pointer to a stack of (up_ref'ed) X509 certificates containing:
+ *      - the EE certificate given in the function arguments (cert)
+ *      - all intermediate certificates up the chain toward the trust anchor
+ *        whereas the (self-signed) trust anchor is not included
+ * returns NULL on error
+ */
+STACK_OF(X509) *ossl_cmp_build_cert_chain(STACK_OF(X509) *certs, X509 *cert)
+{
+    STACK_OF(X509) *chain = NULL, *result = NULL;
+    X509_STORE *store = X509_STORE_new();
+    X509_STORE_CTX *csc = NULL;
+
+    if (certs == NULL || cert == NULL || store == NULL) {
+        CMPerr(0, CMP_R_NULL_ARGUMENT);
+        goto err;
+    }
+
+    csc = X509_STORE_CTX_new();
+    if (csc == NULL)
+        goto err;
+
+    ossl_cmp_X509_STORE_add1_certs(store, certs, 0);
+    if (!X509_STORE_CTX_init(csc, store, cert, NULL))
+        goto err;
+
+    (void)ERR_set_mark();
+    /*
+     * ignore return value as it would fail without trust anchor given in store
+     */
+    (void)X509_verify_cert(csc);
+
+    /* don't leave any new errors in the queue */
+    (void)ERR_pop_to_mark();
+
+    chain = X509_STORE_CTX_get0_chain(csc);
+
+    /* result list to store the up_ref'ed not self-signed certificates */
+    if ((result = sk_X509_new_null()) == NULL)
+        goto err;
+    ossl_cmp_sk_X509_add1_certs(result, chain, 1 /* no self-signed */,
+                                1 /* no duplicates */, 0);
+
+ err:
+    X509_STORE_free(store);
+    X509_STORE_CTX_free(csc);
+    return result;
+}
+
+int ossl_cmp_asn1_octet_string_set1(ASN1_OCTET_STRING **tgt,
+                                    const ASN1_OCTET_STRING *src)
+{
+    if (tgt == NULL) {
+        CMPerr(0, CMP_R_NULL_ARGUMENT);
+        return 0;
+    }
+    if (*tgt == src) /* self-assignment */
+        return 1;
+    ASN1_OCTET_STRING_free(*tgt);
+
+    if (src != NULL) {
+        if ((*tgt = ASN1_OCTET_STRING_dup(src)) == NULL)
+            return 0;
+    } else {
+        *tgt = NULL;
+    }
+
+    return 1;
+}
+
+int ossl_cmp_asn1_octet_string_set1_bytes(ASN1_OCTET_STRING **tgt,
+                                          const unsigned char *bytes, int len)
+{
+    ASN1_OCTET_STRING *new = NULL;
+    int res;
+
+    if (tgt == NULL) {
+        CMPerr(0, CMP_R_NULL_ARGUMENT);
+        return 0;
+    }
+    if (bytes != NULL) {
+        if ((new = ASN1_OCTET_STRING_new()) == NULL
+                || !(ASN1_OCTET_STRING_set(new, bytes, len))) {
+            ASN1_OCTET_STRING_free(new);
+            return 0;
+        }
+    }
+
+    res = ossl_cmp_asn1_octet_string_set1(tgt, new);
+    ASN1_OCTET_STRING_free(new);
```

Comment:
BTW: since this is used to set a secret string value when called from OSSL_CMP_CTX_set1_secretValue
that is even an error to free the temporary object without cleanse.

Reply:
Very good thought. Yet the previous `secretValue` field is already guaranteed to be cleansed by `OSSL_CMP_CTX_set1_secretValue()`.

Reply:
Yes, I saw that and it is OK.